File: qovis_backend/trans/rule/trans_link.py
from enum import Enum
from typing import Optional, Union

from trans.plan.plan_node import PlanNode


class TransLink:
    class LinkKind(Enum):
        EQ = 'eq'
        CH = 'ch'
        PEQ = 'peq'

    UNCHANGED_STEP = ''
    GUESS_STEP = '?'

    @staticmethod
    def mk_eq(node0: PlanNode, node1: PlanNode,
              param0: str, param1: str,
              param_idx0: int = 0, param_idx1: int = 0,
              rules: Optional[list[Union[str, 'Rule']]] = None) -> 'TransLink':
        return TransLink(node0, node1, TransLink.LinkKind.EQ, param0, param1, param_idx0, param_idx1, rules)

    @staticmethod
    def mk_ch(node0: PlanNode, node1: PlanNode,
              param0: str, param1: str,
              param_idx0: int = 0, param_idx1: int = 0,
              rules: Optional[list[Union[str, 'Rule']]] = None) -> 'TransLink':
        return TransLink(node0, node1, TransLink.LinkKind.CH, param0, param1, param_idx0, param_idx1, rules)

    @staticmethod
    def mk_peq(node0: PlanNode, node1: PlanNode,
               param0: str, param1: str,
               param_idx0: int = 0, param_idx1: int = 0,
               rules: Optional[list[Union[str, 'Rule']]] = None) -> 'TransLink':
        return TransLink(node0, node1, TransLink.LinkKind.PEQ, param0, param1, param_idx0, param_idx1, rules)

    def __init__(self, node0: PlanNode, node1: PlanNode, kind: LinkKind,
                 param0: str, param1: str,
                 param_idx0: int, param_idx1: int,
                 rules: Optional[list[Union[str, 'Rule']]] = None):
        self.node0 = node0
        self.node1 = node1
        self.kind = kind
        self.param0 = param0
        self.param1 = param1
        self.param_idx0 = param_idx0
        self.param_idx1 = param_idx1
        self.rules = rules or []

    def merge(self, other: 'TransLink') -> 'TransLink':
        """
        Merge two links into one.
              EQ  CH  PEQ
        EQ    EQ  CH  PEQ
        CH    CH  CH  PEQ
        PEQ  PEQ PEQ  PEQ
        """
        assert self.node1.vid == other.node0.vid
        # process PEQ
        if self.kind == self.LinkKind.PEQ or other.kind == self.LinkKind.PEQ:
            return TransLink.mk_peq(self.node0, other.node1,
                                    self.param0, other.param1,
                                    self.param_idx0, other.param_idx1,
                                    self.rules + other.rules)
        # process EQ and CH
        if self.kind == self.LinkKind.EQ and other.kind == self.LinkKind.EQ:
            return TransLink.mk_eq(self.node0, other.node1,
                                   self.param0, other.param1,
                                   self.param_idx0, other.param_idx1,
                                   self.rules + other.rules)
        # elif other.kind == self.LinkKind.EQ:
        #     return TransLink.mk_ch(self.node0, other.node1,
        #                            self.param0, self.param1,
        #                            self.param_idx0, self.param_idx1)
        # elif self.kind == self.LinkKind.EQ:
        #     return TransLink.mk_ch(self.node0, other.node1,
        #                            other.param0, other.param1,
        #                            other.param_idx0, other.param_idx1)
        else:
            return TransLink.mk_ch(self.node0, other.node1,
                                   self.param0, other.param1,
                                   self.param_idx0, other.param_idx1,
                                   self.rules + other.rules)

    def __eq__(self, other: 'TransLink'):
        return self.node0 is other.node0 and \
               self.node1 is other.node1 and \
               self.kind == other.kind and \
               self.param0 == other.param0 and \
               self.param1 == other.param1 and \
               self.param_idx0 == other.param_idx0 and \
               self.param_idx1 == other.param_idx1

    def __str__(self):
        def node_str(node: PlanNode):
            return f'{node.name}#{node.vid}'

        def param_str(node, param, param_idx):
            return f'{node_str(node)}.{param}[{param_idx}]'

        if self.kind == self.LinkKind.EQ:
            return f'EqLink({param_str(self.node0, self.param0, self.param_idx0)} == ' \
                   f'{param_str(self.node1, self.param1, self.param_idx1)})'
        elif self.kind == self.LinkKind.CH:
            return f'ChLink({param_str(self.node0, self.param0, self.param_idx0)} -> ' \
                   f'{param_str(self.node1, self.param1, self.param_idx1)})'
        elif self.kind == self.LinkKind.PEQ:
            return f'PEqLink({param_str(self.node0, self.param0, self.param_idx0)} == ' \
                   f'{param_str(self.node1, self.param1, self.param_idx1)})'
        else:
            raise Exception(f'Unknown link kind {self.kind}')

    def dump(self):
        return {
            'vid0': self.node0.vid,
            'vid1': self.node1.vid,
            'kind': self.kind.value,
            'p0': self.param0,
            'p1': self.param1,
            'idx0': self.param_idx0,
            'idx1': self.param_idx1,
            'rules': [r if isinstance(r, str) else r.name for r in self.rules],
        }

File: qovis_frontend/src/js/trace/TransLink.js
import {VisTransLink} from "@/js/trace/VisTransLink"

/**
 * @enum {string}
 * @readonly
 */
export const LinkKind = {
  EQ: 'eq',
  CH: 'ch',
  PEQ: 'peq',
}

export class TransLink {
  /** @type {import('@/js/trace/PlanNode').PlanNode} */
  node0
  /** @type {import('@/js/trace/PlanNode').PlanNode} */
  node1
  /** @type {LinkKind} */
  kind
  /** @type {string | null} */
  param0
  /** @type {string | null} */
  param1
  /** @type {int | null} */
  paramIdx0
  /** @type {int | null} */
  paramIdx1
  /** @type {string[]} */
  rules

  /** @type {ParamItem} */
  srcParamItem = null
  /** @type {ParamItem} */
  dstParamItem = null

  /** @type {VisTransLink} */
  visObj = null

  /**
   * @param {PlanNode} node0
   * @param {PlanNode} node1
   * @param {LinkKind} kind
   * @param {string | null} param0
   * @param {string | null} param1
   * @param {int | null} paramIdx0
   * @param {int | null} paramIdx1
   * @param {string[]} rules
   */
  constructor(node0, node1, kind,
              param0, param1, paramIdx0, paramIdx1,
              rules) {
    this.node0 = node0
    this.node1 = node1
    this.kind = kind
    this.param0 = param0
    this.param1 = param1
    this.paramIdx0 = paramIdx0
    this.paramIdx1 = paramIdx1
    this.rules = rules

    this.srcParamItem = this.node0.paramItemsMap.get(param0)[paramIdx0 ?? 0]
    this.dstParamItem = this.node1.paramItemsMap.get(param1)[paramIdx1 ?? 0]
  }

  /**
   * @returns {VisTransLink}
   */
  createVisObj() {
    this.visObj = new VisTransLink(this)
    return this.visObj
  }

  toString() {
    // python code
    //     def __str__(self):
    //         def node_str(node: PlanNode):
    //             return f'{node.name}#{node.vid}'
    //
    //         def param_str(node, param, param_idx):
    //             return f'{node_str(node)}.{param}[{param_idx}]'
    //
    //         if self.kind == self.LinkKind.EQ:
    //             return f'EqLink({param_str(self.node0, self.param0, self.param_idx0)} == ' \
    //                    f'{param_str(self.node1, self.param1, self.param_idx1)})'
    //         elif self.kind == self.LinkKind.CH:
    //             return f'ChLink({param_str(self.node0, self.param0, self.param_idx0)} -> ' \
    //                    f'{param_str(self.node1, self.param1, self.param_idx1)})'
    //         elif self.kind == self.LinkKind.PEQ:
    //             return f'PEqLink({param_str(self.node0, self.param0, self.param_idx0)} == ' \
    //                    f'{param_str(self.node1, self.param1, self.param_idx1)})'
    //         else:
    //             raise Exception(f'Unknown link kind {self.kind}')

    const nodeStr = (node) => `Plan#${node.plan.pid}.${node.name}#${node.vid}`
    const paramStr = (node, param, paramIdx) => `${nodeStr(node)}.${param}[${paramIdx}]`

    if (this.kind === LinkKind.EQ) {
      return `EqLink(${paramStr(this.node0, this.param0, this.paramIdx0)} == ` +
          `${paramStr(this.node1, this.param1, this.paramIdx1)})`
    }
    if (this.kind === LinkKind.CH) {
      return `ChLink(${paramStr(this.node0, this.param0, this.paramIdx0)} -> ` +
          `${paramStr(this.node1, this.param1, this.paramIdx1)})`
    }
    if (this.kind === LinkKind.PEQ) {
      return `PEqLink(${paramStr(this.node0, this.param0, this.paramIdx0)} == ` +
          `${paramStr(this.node1, this.param1, this.paramIdx1)})`
    }
    throw new Error(`Unknown link kind ${this.kind}`)
  }

  // /**
  //  * @param {TransLink} other
  //  * @return {TransLink}
  //  */
  // merge(other) {
  //   // python code
  //   //         assert self.node1.vid == other.node0.vid
  //   //         assert self.kind != self.LinkKind.PEQ and other.kind != self.LinkKind.PEQ
  //   //         if self.kind == self.LinkKind.EQ and other.kind == self.LinkKind.EQ:
  //   //             return TransLink.mk_eq(self.node0, other.node1,
  //   //                                    self.param0, other.param1,
  //   //                                    self.param_idx0, other.param_idx1)
  //   //         elif other.kind == self.LinkKind.EQ:
  //   //             return TransLink.mk_ch(self.node0, other.node1,
  //   //                                    self.param0, self.param1,
  //   //                                    self.param_idx0, self.param_idx1)
  //   //         elif self.kind == self.LinkKind.EQ:
  //   //             return TransLink.mk_ch(self.node0, other.node1,
  //   //                                    other.param0, other.param1,
  //   //                                    other.param_idx0, other.param_idx1)
  //   //         else:
  //   //             return TransLink.mk_ch(self.node0, other.node1,
  //   //                                    self.param0, other.param1,
  //   //                                    self.param_idx0, other.param_idx1)
  //   if (this.node1.vid !== other.node0.vid) {
  //     throw new Error('cannot merge two links')
  //   }
  //   if (this.kind === LinkKind.PEQ || other.kind === LinkKind.PEQ) {
  //     throw new Error('cannot merge two links')
  //   }
  //   if (this.kind === LinkKind.EQ && other.kind === LinkKind.EQ) {
  //     return new TransLink(
  //         this.node0, other.node1,
  //         LinkKind.EQ,
  //         this.param0, other.param1,
  //         this.paramIdx0, other.paramIdx1)
  //   } else if (other.kind === LinkKind.EQ) {
  //     return new TransLink(
  //         this.node0, other.node1,
  //         LinkKind.CH,
  //         this.param0, this.param1,
  //         this.paramIdx0, this.paramIdx1)
  //   } else if (this.kind === LinkKind.EQ) {
  //     return new TransLink(
  //         this.node0, other.node1,
  //         LinkKind.CH,
  //         other.param0, other.param1,
  //         other.paramIdx0, other.paramIdx1)
  //   } else {
  //     return new TransLink(
  //         this.node0, other.node1,
  //         LinkKind.CH,
  //         this.param0, other.param1,
  //         this.paramIdx0, other.paramIdx1)
  //   }
  // }
}
